tpe/lib/src/Model.cc

```diff
@@ -51,14 +51,42 @@ Entity &Model::AddLink()
   return *it->second.get();
 }
 
+//////////////////////////////////////////////////
+Entity &Model::AddModel()
+{
+  std::size_t modelId = Entity::GetNextId();
+  const auto[it, success]  = this->GetChildren().insert(
+      {modelId, std::make_shared<Model>(modelId)});
+
+  it->second->SetParent(this);
+  this->ChildrenChanged();
+  return *it->second.get();
+}
+
 //////////////////////////////////////////////////
 Entity &Model::GetCanonicalLink()
 {
   // return first link as canonical link
-  return *this->GetChildren().begin()->second;
+  for (auto &it : this->GetChildren())
```

Comment:
this logic doesn't match the approach in https://github.com/osrf/sdformat/pull/341 (which is still under review), which chooses the first link if a model has links, and otherwise gets the canonical link of the first nested model

* https://github.com/osrf/sdformat/pull/341/files#diff-8d3009e318ee9ebbc70c49f879cf8b57R630-R632

Reply:
thanks for pointing that out. I've updated the logic and added test in 855b671